From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NickAc <32451103+NickAcPT@users.noreply.github.com>
Date: Fri, 25 Jun 2021 09:27:44 +0100
Subject: [PATCH] Custom enchantments API


diff --git a/src/main/java/io/github/nickacpt/deepslate/enchantments/CustomEnchantment.java b/src/main/java/io/github/nickacpt/deepslate/enchantments/CustomEnchantment.java
new file mode 100644
index 0000000000000000000000000000000000000000..1e90ba10a567ca4d449866ec556999fe842080e7
--- /dev/null
+++ b/src/main/java/io/github/nickacpt/deepslate/enchantments/CustomEnchantment.java
@@ -0,0 +1,100 @@
+package io.github.nickacpt.deepslate.enchantments;
+
+import java.util.EnumSet;
+import java.util.Locale;
+import java.util.Set;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import org.bukkit.Keyed;
+import org.bukkit.NamespacedKey;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.entity.EntityCategory;
+import org.bukkit.entity.Player;
+import org.bukkit.event.block.BlockBreakEvent;
+import org.bukkit.inventory.EquipmentSlot;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Plugin provided custom enchantment
+ */
+public abstract class CustomEnchantment extends Enchantment implements Keyed {
+
+    protected CustomEnchantment(@NotNull NamespacedKey key) {
+        super(key);
+    }
+
+    @Override
+    public @NotNull String getName() {
+        return "CUSTOM_" + getKey().getNamespace().toUpperCase(Locale.ROOT) + "_" + getKey().getKey().toUpperCase(Locale.ROOT);
+    }
+
+    @Override
+    public boolean isCursed() {
+        return false;
+    }
+
+    @Override
+    public boolean isTreasure() {
+        return false;
+    }
+
+    @Override
+    public boolean conflictsWith(@NotNull Enchantment other) {
+        return false;
+    }
+
+    @Override
+    public float getDamageIncrease(int level, @NotNull EntityCategory entityCategory) {
+        return 0;
+    }
+
+    @Override
+    public boolean canEnchantItem(@NotNull ItemStack item) {
+        return true;
+    }
+
+    @Override
+    public boolean isDiscoverable() {
+        return true;
+    }
+
+    @Override
+    public boolean isTradeable() {
+        return true;
+    }
+
+    @NotNull
+    @Override
+    public Set<EquipmentSlot> getActiveSlots() {
+        return EnumSet.allOf(EquipmentSlot.class);
+    }
+
+    @NotNull
+    public abstract Component displayName();
+
+    @NotNull
+    @Override
+    public Component displayName(int level) {
+        Component component = displayName();
+
+        if (level != 1 || this.getMaxLevel() != 1) {
+            component = component.append(Component.space()).append(Component.translatable("enchantment.level." + level));
+        }
+
+        if (this.isCursed()) {
+            component = component.color(NamedTextColor.RED);
+        } else {
+            component = component.color(NamedTextColor.GRAY);
+        }
+
+        return component;
+    }
+
+    public void onBlockBreak(@NotNull BlockBreakEvent event, int level, @NotNull Player player, @NotNull ItemStack itemStack) {}
+
+    @Override
+    public @NotNull String translationKey() {
+        return getName();
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 390a0d09b78e46cfebe78bfa3d1dd73aedca8256..070588251cfd203db4ab469394ecf4fc704a1da4 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2102,4 +2102,14 @@ public final class Bukkit {
         server.removeFuel(material);
     }
     // Purpur end
+
+    // Deepslate start
+    /**
+     * Register a plugin provided custom enchantment
+     * @param enchantment The enchantment to register
+     */
+    public static void registerEnchantment(@NotNull io.github.nickacpt.deepslate.enchantments.CustomEnchantment enchantment) {
+        server.registerEnchantment(enchantment);
+    }
+    // Deepslate end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 88767865cc2dce0e341eef0f53323af6f95fb841..868f7dd5fafd95306fd7bd8a5981b46065197cde 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1835,4 +1835,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public void removeFuel(@NotNull Material material);
     // Purpur end
+
+    // Deepslate start
+
+    /**
+     * Register a plugin provided custom enchantment
+     * @param enchantment The enchantment to register
+     */
+    void registerEnchantment(@NotNull io.github.nickacpt.deepslate.enchantments.CustomEnchantment enchantment);
+    // Deepslate end
 }
diff --git a/src/main/java/org/bukkit/enchantments/Enchantment.java b/src/main/java/org/bukkit/enchantments/Enchantment.java
index 1e1c5a9d9a769018c4604e6e44fc5ed2312981e9..f835130f9ed452446e1ab01125999a8f64d9f03c 100644
--- a/src/main/java/org/bukkit/enchantments/Enchantment.java
+++ b/src/main/java/org/bukkit/enchantments/Enchantment.java
@@ -379,7 +379,7 @@ public abstract class Enchantment implements Keyed, net.kyori.adventure.translat
     public static void registerEnchantment(@NotNull Enchantment enchantment) {
         if (byKey.containsKey(enchantment.key) || byName.containsKey(enchantment.getName())) {
             throw new IllegalArgumentException("Cannot set already-set enchantment");
-        } else if (!isAcceptingRegistrations()) {
+        } else if (!(enchantment instanceof io.github.nickacpt.deepslate.enchantments.CustomEnchantment) && !isAcceptingRegistrations()) { // Deepslate - Always accept registration of custom enchantments
             throw new IllegalStateException("No longer accepting new enchantments (can only be done by the server implementation)");
         }
 
