From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NickAc <32451103+NickAcPT@users.noreply.github.com>
Date: Sat, 24 Jul 2021 22:41:17 +0100
Subject: [PATCH] Fix GameProfile conversion error


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
index f61c313195c3d16d996721b2f8cd0d9a10ce1aaf..d52a7539f95214ab715871777a0b36c3d8c8ae52 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -152,6 +152,7 @@ public class SkullBlockEntity extends BlockEntity {
             profileCache.getAsync(owner.getName(), (profile) -> {
                 Util.PROFILE_EXECUTOR.execute(() -> { // Paper - not a good idea to use BLOCKING OPERATIONS on the worldgen executor
                     Util.ifElse(profile, (profilex) -> {
+                        if (profilex == null) return; // Deepslate - fix conversion error
                         Property property = Iterables.getFirst(profilex.getProperties().get("textures"), (Property)null);
                         if (property == null) {
                             profilex = sessionService.fillProfileProperties(profilex, true);
