From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NickAc <32451103+NickAcPT@users.noreply.github.com>
Date: Sat, 24 Jul 2021 22:41:17 +0100
Subject: [PATCH] Fix GameProfile conversion error


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
index 6381544b0038de9a09c01238638e4e127e4eddc6..2f3606b08ff50254ff2fe1dd92ab43c3c0a0fc3d 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -152,6 +152,7 @@ public class SkullBlockEntity extends BlockEntity {
             profileCache.getAsync(owner.getName(), (profile) -> {
                 Util.backgroundExecutor().execute(() -> {
                     Util.ifElse(profile, (profilex) -> {
+                        if (profilex == null) return; // Deepslate - fix conversion error
                         Property property = Iterables.getFirst(profilex.getProperties().get("textures"), (Property)null);
                         if (property == null) {
                             profilex = sessionService.fillProfileProperties(profilex, true);
